net/HttpRequest.hpp

```diff
@@ -1623,24 +1657,35 @@ class Session final : public ProtocolHandlerInterface
         {
             LOG_TRC("onHandshakeFail");
             _handshakeSslVerifyFailure = socket->getSslVerifyResult();
+            _result = net::asyncConnectResult::SSLHandShakeFailure;
         }
+
+        callOnConnectFail();
     }
 
     void onDisconnect() override
     {
+
+        bool completed = false;
         // Make sure the socket is disconnected and released.
         std::shared_ptr<StreamSocket> socket = _socket.lock();
         if (socket)
         {
+            completed = socket->getOutBuffer().empty() && socket->getInBuffer().empty();
```

Comment:
I am using this heuristic `outbuffer.empty() && inBuffer.empty()` to determine if the request was completed, this seems brittle, suggestion welcome to improve on those.

Reply:
Very skeptical. Can @Ashod or @caolanm look over this.

My expectation is that we don't need to deeply change the async socket code - to add complexity here. Possibly we need to add a more detailed way of fetching errors on socket connections - I can believe a single new virtual function to do that and perhaps a member to store the last error or whatever might make sense as a minimal change - but not read it. 

Separately  I would  be very surprised if an empty outgoing buffer meant anything at all. It just means some combination of kernel and/or underlying SSL layer was able to take the data.

Please adapt to whatever @Ashod recommends =)

Reply:
To add more context: currently the code will mark any request's response `_state` to  `Error` status in `Response::finish` by way of `HttpRequest::onDisconnect` when the socket is closed in `Socket::handlePoll`.

But most of the time a disconnection is the normal/complete case at the end of the request.
`events & POLLHUP` might be more appropriate and we'd need to differentiate the two cases of onDisconnect.

Reply:
HTTP/1.1 wants optionally a "Connection: close" header to signal the connection/socket will be closed.  https://datatracker.ietf.org/doc/html/rfc2616#section-14.10
We interpret this by calling the `onDisconnect()` callback in `newRequest` that will mark the response in error with `_response->finish`.

I have been testing with a HTTP/1 server `openssl s_server` and that ends up, calling `onDisconnect` from `StreamSocket::handlePoll`

Reply:
This is a minor concern that shouldn't affect much the feature and can be fixed separately.